File: ip23/gibbs.m
% A simple Gibbs sampler algorithm

clear all; close all;
p = @(x,y) exp(-10*(x.^2-y).^2-(y-1/4).^4);
x = [0,0];
nsamples = 5000;
samples = [];
xx = linspace(-2,2,10000);
for jj = 2:nsamples
    u = rand;
    P = cumsum(p(xx,x(2)));
    P = P/P(end);
    ind = find(u <= P,1,'first');
    x(1) = xx(ind);
    u = rand;
    P = cumsum(p(x(1),xx));
    P = P/P(end);
    ind = find(u <= P,1,'first');
    x(2) = xx(ind);
    samples = [samples;x];
end
% At this point, one would usually discard some number of
% the initial samples (the burn-in period). Here, we omit
% this step.
[X,Y] = meshgrid(linspace(-2,2));
figure;
contour(X,Y,p(X,Y)), hold on;
plot(samples(:,1),samples(:,2),'k.')
title(['Single component Gibbs sampler with ',num2str(nsamples),' samples'])
hold off;

figure;
subplot(2,1,1)
plot(samples(:,1));
xlabel('Sample history of x_1');
ylabel('x_1');
hold on;
subplot(2,1,2)
plot(samples(:,2))
xlabel('Sample history of x_2');
ylabel('x_2');

% Note: autocovariances are computed after burn-in is removed!
figure;
ac1 = autocovariance(samples(:,1));
ac2 = autocovariance(samples(:,2));
N_ac = 100;
plot(0:N_ac,ac1(1:N_ac+1),'ro',0:N_ac,ac2(1:N_ac+1),'bo')
pbaspect([1 1 1])
legend('horizontal component','vertical component')

File: s4ds_vara_2023/gibbs.py
# A simple Gibbs sampler algorithm
import numpy as np
import matplotlib.pyplot as plt

p = lambda x,y: np.exp(-10*(x**2-y)**2-(y-1/4)**4)
x = np.array([0,0])
nsamples = 5000
samples = []
xx = np.linspace(-2,2,10000)
for jj in range(1,nsamples):
    u = np.random.uniform()
    P = np.cumsum(p(xx,x[1]))
    P = P/P[-1]
    ind = np.where(u<=P)[0][0]
    x1 = xx[ind]
    u = np.random.uniform()
    P = np.cumsum(p(x1,xx))
    P = P/P[-1]
    ind = np.where(u<=P)[0][0]
    x2 = xx[ind]
    x = np.array([x1,x2])
    samples.append(x)
# At this point, one would usually discard some number of
# the initial samples (the burn-in period). Here, we omit
# this step.
samples = list(zip(*samples)) # transpose list
X,Y = np.meshgrid(np.linspace(-2,2),np.linspace(-2,2))
plt.contour(X,Y,p(X,Y))
plt.plot(samples[0],samples[1],'.',color='black',markersize=2)
plt.title('Single component Gibbs sampler with ' + str(nsamples) + ' samples')
plt.show()

fig,ax = plt.subplots(2,1)
ax[0].plot(samples[0],linewidth=.5)
ax[0].set_xlabel('Sample history of $x_1$')
ax[0].xaxis.set_ticks_position('top')
ax[0].set_ylabel('$x_1$')
ax[1].plot(samples[1],linewidth=.5)
ax[1].set_xlabel('Sample history of $x_2$')
ax[1].set_ylabel('$x_2$')
plt.show()

# Function to compute the autocovariance
def autocovariance(f):
    N = len(f)
    gamma2 = np.zeros(N-1)
    f_c = f-np.mean(f)
    gamma2_0 = np.mean(f_c**2)
    for k in np.arange(N-1,0,-1):
        jj = np.arange(1,N-k+1)
        gamma2[k-1] = 1/(gamma2_0*(N-k)) * np.sum(f_c[jj-1]*f_c[jj+k-1])
    return gamma2

# Note: autocovariances are computed after burn-in is removed!
ac1 = autocovariance(samples[0])
ac2 = autocovariance(samples[1])
N_ac = 100
plt.plot(range(0,N_ac+1),ac1[0:N_ac+1],'o',markerfacecolor='none',markeredgecolor='red',label='horizontal component')
plt.plot(range(0,N_ac+1),ac2[0:N_ac+1],'o',markerfacecolor='none',markeredgecolor='blue',label='vertical component')
plt.legend()
plt.show()